coursier.bzl

```diff
@@ -480,6 +480,24 @@ def _windows_check(repository_ctx):
                  "This is typically `c:\\msys64\\usr\\bin\\bash.exe`. For more information, read " +
                  "https://docs.bazel.build/versions/master/install-windows.html#getting-bazel")
 
+# Deterministically compute a signature of a list of artifacts in the dependency tree.
+# This is to prevent users to manually editing maven_install.json.
+def _compute_dependency_tree_signature(artifacts):
+    # A collection of elements from the dependency tree to be sorted and hashed
+    # into a signature for maven_install.json.
+    signature_inputs = []
+    for artifact in artifacts:
+        signature_inputs.append(artifact["coord"])
+        if artifact["file"] != None:
+            signature_inputs.extend([
+                artifact["sha256"],
+                artifact["file"],
+                artifact["url"]
+            ])
+        if len(artifact["dependencies"]) > 0:
+            signature_inputs.append(",".join(sorted(artifact["dependencies"])))
+    return hash(repr(sorted(signature_inputs)))
```

Comment:
I worry that sorting an entire `signature_inputs` could be slow when dealing with a huge `maven_install.json`, especially it seems like we will compute the hash for every `bazel build //...`. In addition, I found that
```
{
    "coord": "artifact1",
    "dependencies": [
        "deps1"
    ]
},
{
    "coord": "artifact2",
    "dependencies": [
        "deps2"
    ]
}
```
and 
```
{
    "coord": "artifact1",
    "dependencies": [
        "deps2"
    ]
},
{
    "coord": "artifact2",
    "dependencies": [
        "deps1"
    ]
}
```
could yield to the same hash. Even though users are very unlikely to make this kind of change, but the algorithm can be more robust.

I am thinking of something like
```
signature_inputs = []
for artifact in artifacts:
    artifact_group = []
    artifact_group.append(artifact["coord"])
    if artifact["file"] != None:
        artifact_group.extend([
            artifact["sha256"],
            artifact["file"],
            artifact["url"]
        ])
    if len(artifact["dependencies"]) > 0:
        artifact_group.append(",".join(sorted(artifact["dependencies"])))
    signature_inputs.append(artifact_group)
return hash(repr(sorted(signature_inputs, key=lambda arti: arti[0])))
```
It ties the corresponding `sha256`, `file`, `url`, and `dependencies` to the `coord`, and only sorts the entire artifact group.

Reply:
From my experience developing this so far, performance hasn't been an issue even with 2000+ lines in regression_testing_install.json. 

Thanks for the excellent suggestion to the algorithm, PTAL (there isn't `lambda` in Starlark, and I don't think there's a need to sort on `coord`).

Reply:
> From my experience developing this so far, performance hasn't been an issue even with 2000+ lines in regression_testing_install.json.

We currently have up to 15000 lines, but we could worry about the performance once it is really a concern.

> there isn't lambda in Starlark, and I don't think there's a need to sort on coord

Good point, sorry I almost forgot it's not Python.